libraries/chain/include/eosio/chain/webassembly/preconditions.hpp

```diff
@@ -71,21 +84,12 @@ namespace eosio { namespace chain { namespace webassembly {
       return f128_is_nan( f );
    }
 
-   EOS_VM_PRECONDITION(early_validate_pointers,
+   EOS_VM_PRECONDITION(null_pointer_check,
          EOS_VM_INVOKE_ON_ALL([&](auto&& arg, auto&&... rest) {
             using namespace eosio::vm;
             using arg_t = std::decay_t<decltype(arg)>;
-            if constexpr (is_reference_proxy_type_v<arg_t>) {
-               if constexpr (is_span_type_v<dependent_type_t<arg_t>>) {
-                  using dep_t = dependent_type_t<arg_t>;
-                  const dep_t& s = arg;
-                  EOS_ASSERT( s.size() <= std::numeric_limits<wasm_size_t>::max() / (wasm_size_t)sizeof(dependent_type_t<dep_t>),
-                        wasm_execution_error, "length will overflow" );
-                  volatile auto check = *(reinterpret_cast<const char*>(arg.original_ptr) + s.size_bytes() - 1);
-                  ignore_unused_variable_warning(check);
-               } else {
-                  EOS_ASSERT(arg.original_ptr != ctx.get_interface().get_memory(), wasm_execution_error, "references cannot be created for null pointers");
-               }
+            if constexpr (decltype(detail::is_legacy_ptr(std::declval<arg_t>()))::value && !is_span_type_v<dependent_type_t<arg_t>>) {
```

Comment:
I don’t agree with this one fully either. The reason it is in eos-vm is because we already have multiple targets that will utilize that type of copying back functionality. If we move it out specifically for eos, then we need to create and maintain separate ones for the other projects. The reference_proxy is not just for the legacy stuff, it’s also for pointers and references. If you’d like a separate pointer_proxy in eos, that is fine. I don’t see any harm in that.

Reply:
Good point on others benefiting from an array copying wrapper. 

How about the following?

```cpp
template <typename T, std::size_t LegacyAlign=0>
struct reference_proxy;
```

becomes

```cpp
template <typename T, std::size_t Extent = dynamic_extent, std::size_t LegacyAlign=0>
struct proxy_span;
```

where we can again, for now, only support `Extent == dynamic_extent` or `Extent == 1`. 

A `proxy_span<T, 1, LegacyAlign>` would be a specialization of the template that has a similar structure and behavior as `reference_proxy<T, LegacyAlign>` where `T` is not a `span`. EOS VM would know to expand this out as a `void*` within the host function that WASM code calls.

 A `proxy_span<T, dynamic_extent, LegacyAlign>` would behave like a `reference_proxy<span<T>, LegacyAlign>`. EOS VM would know to expand this out as `void*, wasm_size_t` withn the host function that WASM code calls.

The `proxy_span` would only be able to implicitly convert to the appropriate `span` type (no implicit conversion to a pointer or to a reference) and it would not have a `ref()` member function.

In EOSIO, `legacy_ptr<T>` would be a type alias to `eosio::vm::proxy_span<T, 1, alignof(T)>` and `legacy_array_ptr<T>` (which I still think should be renamed to `legacy_span`) would be a type alias to `eosio::vm::proxy_span<T, dynamic_extent, alignof(T)>`.

And then the condition in the `if constexpr` here gets to be simplified to just `decltype(detail::is_legacy_ptr(std::declval<arg_t>()))::value`.

Reply:
Currently `reference_proxy<T*>` and `reference_proxy<T&>` are both equivalent to `reference_proxy<T>`, which is strange.  I think we need a total of three templates.  reference_proxy, pointer_proxy, and span_proxy.  We can either make them separate templates or three specializations of reference_proxy (and leave the primary template undefined).